tools/xdp/s2n-quic-xdp/src/if_xdp.rs

```diff
@@ -0,0 +1,265 @@
+}
+
+impl Default for Address {
+    fn default() -> Self {
+        Self {
+            family: libc::PF_XDP as _,
+            flags: Default::default(),
+            ifindex: 0,
+            queue_id: 0,
+            shared_umem_fd: 0,
+        }
+    }
+}
+
+impl Address {
+    /// Resolves the `ifindex` for the provided `if_name` and associates it with the address.
+    ///
+    /// If the device does not exist, then an error is returned.
+    #[inline]
+    pub fn set_if_name(&mut self, name: &CStr) -> Result<&mut Self> {
+        unsafe {
+            let ifindex = libc::if_nametoindex(name.as_ptr());
+
+            // https://man7.org/linux/man-pages/man3/if_nametoindex.3.html
+            // > On success, if_nametoindex() returns the index number of the
+            // > network interface; on error, 0 is returned and errno is set to
+            // > indicate the error.
+            if ifindex == 0 {
+                return Err(std::io::Error::last_os_error());
+            }
+
+            self.ifindex = ifindex;
+        }
+        Ok(self)
+    }
+}
+
+bitflags!(
+    /// Flags set on a particular ring state `flags` field
+    ///
+    /// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L41
+    #[derive(Default)]
+    #[repr(transparent)]
+    pub struct RingFlags: u32 {
+        const NEED_WAKEUP = 1 << 0;
+    }
+);
+
+/// A structure to communicate the offsets for an individual ring for an AF_XDP socket
+///
+/// Note that the `flags` field was added in kernel 5.4.
+///
+/// See https://github.com/torvalds/linux/blob/9116e5e2b1fff71dce501d971e86a3695acc3dba/include/uapi/linux/if_xdp.h#L28-L32
+#[derive(Clone, Copy, Debug)]
+#[repr(C)]
+pub struct RingOffsetV1 {
+    pub producer: u64,
+    pub consumer: u64,
+    pub desc: u64,
+}
+
+/// A structure to communicate the offsets for an individual ring for an AF_XDP socket
+///
+/// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L43
+#[derive(Clone, Copy, Debug, Default)]
+#[repr(C)]
+pub struct RingOffsetV2 {
+    pub producer: u64,
+    pub consumer: u64,
+    pub desc: u64,
+    pub flags: u64,
+}
+
+impl RingOffsetV2 {
+    /// Converts an offset description from an older version
+    #[inline]
+    fn set_v1(&mut self, v1: RingOffsetV1) {
+        // Logic from https://github.com/xdp-project/xdp-tools/blob/a76e7a2b156b8cfe38992206abe9df1df0a29e38/lib/libxdp/xsk.c#L197
+        self.producer = v1.producer;
+        self.consumer = v1.consumer;
+        self.desc = v1.desc;
+        self.flags = v1.consumer + (size_of::<u32>() as u64);
+    }
+}
+
+/// A structure to communicate the offsets of each of the 4 rings for an AF_XDP socket
+///
+/// See https://github.com/torvalds/linux/blob/2bac7dc169af3cd4a0cb5200aa1f7b89affa042a/include/uapi/linux/if_xdp.h#L50
+#[derive(Clone, Copy, Debug, Default)]
+#[repr(C)]
+pub struct MmapOffsets<O = RingOffsetV2> {
```

Comment:
why is this generic? In the link it seems like all values are only of type RingOffsetV2

Reply:
There's a V1 on this line https://github.com/aws/s2n-quic/blob/82bde64370bbe6c8c10988ef52ba034795d022d6/tools/xdp/s2n-quic-xdp/src/if_xdp.rs#L169